src/eos-updater-util.c

```diff
@@ -154,14 +211,147 @@ eos_updater_resolve_upgrade (EosUpdater  *updater,
+                                          g_free, g_free);
+  if (g_file_test (DT_COMPATIBLE, G_FILE_TEST_EXISTS))
+    { /* ARM */
+      gs_unref_object GFile *fp = NULL;
+      gs_free gchar *fc = NULL;
+
+      fp = g_file_new_for_path (DT_COMPATIBLE);
+      if (g_file_load_contents (fp, NULL, &fc, NULL, NULL, NULL))
+        {
+          gs_strfreev gchar **sv = g_strsplit (fc, ",", -1);
+
+          if (sv && sv[0])
+            g_hash_table_insert (hw_descriptors, g_strdup ("sys_vendor"),
+                                 g_strdup (g_strstrip (sv[0])));
+          if (sv && sv[1])
+            g_hash_table_insert (hw_descriptors, g_strdup ("product_name"),
+                                 g_strdup (g_strstrip (sv[1])));
+        }
+    }
+  else
+    { /* X86 */
+      guint i;
+
+      for (i = 0; dmi_attributes[i]; i++)
+        {
+          gs_free gchar *path = NULL;
+          gs_unref_object GFile *fp = NULL;
+          gs_free gchar *fc = NULL;
+          gsize len;
+
+          path = g_strconcat (DMI_PATH, dmi_attributes[i], NULL);
+          fp = g_file_new_for_path (path);
+          if (g_file_load_contents (fp, NULL, &fc, &len, NULL, NULL))
+          {
+              if (len > 128)
+                fc[128] = '\0';
+              g_hash_table_insert (hw_descriptors, g_strdup (dmi_attributes[i]),
+                                   g_strdup (g_strstrip (fc)));
+          }
+        }
+    }
+
+  if (!g_hash_table_lookup (hw_descriptors, "sys_vendor"))
+    g_hash_table_insert (hw_descriptors, g_strdup ("sys_vendor"),
+                         g_strdup ("EOSUNKNOWN"));
+
+  if (!g_hash_table_lookup (hw_descriptors, "product_name"))
+    g_hash_table_insert (hw_descriptors, g_strdup ("product_name"),
+                         g_strdup ("EOSUNKNOWN"));
+
+  vendor = g_strdup (g_hash_table_lookup (hw_descriptors, "sys_vendor"));
+  product = g_strdup (g_hash_table_lookup (hw_descriptors, "product_name"));
+  p_id = g_strconcat (cleanstr (vendor), " ", cleanstr (product), NULL);
+  message ("Product group: %s", p_id);
+
+  /* Build the HTTP URI */
+  g_hash_table_insert (hw_descriptors, g_strdup ("ref"), g_strdup (o_ref));
+  g_hash_table_insert (hw_descriptors, g_strdup ("commit"), g_strdup (booted));
+  query = soup_form_encode_hash (hw_descriptors);
+  uri = g_strconcat (baseurl, "/", BRANCHES_CONFIG_PATH, "?", query, NULL);
+  g_hash_table_destroy (hw_descriptors);
+  message ("Branches configuration URI: %s", uri);
+
+  /* Download branch configuration data */
+  soup = soup_session_new ();
+  msg = soup_message_new ("GET", uri);
+  status = soup_session_send_message (soup, msg);
+  if (!SOUP_STATUS_IS_SUCCESSFUL (status))
+    {
+      g_set_error (error, G_IO_ERROR, G_IO_ERROR_HOST_NOT_FOUND,
+                   "Failed to download branch config data (HTTP %d),"
+                   " cannot upgrade", status);
+      goto out;
+    }
+
+  g_object_get (G_OBJECT (msg), "response-body", &msg_body, NULL);
```

Comment:
There's no need to cast to `G_OBJECT()` because `g_object_get()` takes a `gpointer`. Also, pretty sure you need to free `msg_body` here.